bingo.js

```diff
@@ -602,9 +675,119 @@ function setSquareColor(square, colorClass)
 {
 	ALL_COLOURS.forEach(c => square.removeClass(c));
 	square.addClass(colorClass);
+	updateCurrentSheetProgress();
 }
 
-function copySeedToClipboard(id, event)
+function loadBingoProgress(sheetToLoad)
+{
+	/* 
+	 * Expected input for sheetToLoad is [1,3,2,5,..]
+	 * where index of array is numeric value of square id
+	 * and cell value is index of ALL_COLOURS 
+	 */
+	if (typeof sheetToLoad !== 'undefined' && sheetToLoad.length) {
+		forEachSquare((i, square) => {
+			$.each(ALL_COLOURS, function(index, colorClass){
+	       square.removeClass(colorClass);
+	    });
+			square.addClass(ALL_COLOURS[sheetToLoad[i]]);
+		});
+	}
+}
+
+var sheetProgressDebounce = null;
+function updateCurrentSheetProgress()
+{
+	if(sheetProgressDebounce) clearTimeout(sheetProgressDebounce);
+	sheetProgressDebounce = setTimeout(function() {
+			updateCurrentSheetProgressDebounced()
+	}, 100);
+}
+
+function updateCurrentSheetProgressDebounced()
+{
+	if (! isLocalStorageIsAvailable()) {
+		return;
+	}
+
+	processSheetAndStoreAs(SHEET_PROGRESS_KEY + generateNewUrlParam().replace(/-\d-\d/, ''));
+}
+
+function processSheetAndStoreAs(localStorageKey)
+{
+	var squares = generateSquareProgress();
+	result = {
+		squares: squares,
+		timestamp: Date.now(),
+		haveMarkedSquares: squares.reduce(function(a,b){return !!a | !!b})
+	};
+	
+	// save only if current sheet have marked squares or already in storage
+	if (result.haveMarkedSquares || localStorage.getItem(localStorageKey)) {
+		pushNewLocalSetting(localStorageKey, JSON.stringify(result));
+	}
+	// update &progress param is such exists
+	if (gup(URL_PARAM_PROGRESS)) {
+		pushNewUrl(generateNewUrlParamWithProgress(squares));
+	}
+	checkAndRemoveOldSaves();
+}
+
+function checkAndRemoveOldSaves()
+{
+	var saves = getSaves();
+	if (saves.length > SHEET_PROGRESS_SAVE_LIMIT) {
+		var oldestSave = saves.slice(-1).pop();
+		if (oldestSave) {
+			localStorage.removeItem(oldestSave.key);
+		}
+	}
+}
+
+function getSaves()
+{
+	var saves = [];
+	for (i = 0; i < localStorage.length; i++) {
+		key = localStorage.key(i);
+		if (key.slice(0, SHEET_PROGRESS_KEY.length) === SHEET_PROGRESS_KEY) {
+			saves.push({
+				key: key,
+				data: JSON.parse(localStorage.getItem(key))
+			});
+		}
+	}
+
+	saves.sort(function(a, b) {
+		// Compare the 2 dates
+		if (a.data.timestamp < b.data.timestamp) return 1;
+		if (a.data.timestamp > b.data.timestamp) return -1;
+		return 0;
+	});
+	return saves;
+}
+
+function generateSquareProgress()
+{
+	var squares = []
+	forEachSquare((i, square) => {
+		squares.push(ALL_COLOURS.indexOf(square.attr('class') || ''));
+	});
+	return squares;
+}
+
+function checkRecentCurrentSheet()
+{
+	if (! isLocalStorageIsAvailable() || gup(URL_PARAM_PROGRESS)) {
+		return;
+	}
+
+	var latestSheet = JSON.parse(localStorage.getItem(SHEET_PROGRESS_KEY + generateNewUrlParam().replace(/-\d-\d/, '')));
```

Comment:
`SHEET_PROGRESS_KEY + generateNewUrlParam().replace(/-\d-\d/, ''))` should be extracted into a separate function. It's repeated in `updateCurrentSheetProgressDebounced`.

The code might be clearer, if saving/loading feature generated the keys without using `generateNewUrlParam()`.

Reply:
> The code might be clearer, if saving/loading feature generated the keys without using `generateNewUrlParam()`.

This is just so that the reader doesn't need to check what `generateNewUrlParam()` does when figuring out what this regexp removes from the string.